content/blog/getting-started-with-java-development-2023.asciidoc

```diff
@@ -0,0 +1,211 @@
+Java source code is compiled into class files which then are loaded into the JVM and executed.
+Normally, this is done in two steps: first running the compiler _javac_, then executing the program using the _java_ binary.
+For quick testing and exploring, both steps can be combined, so you can execute your "Hello World" program like this:
+
+[source,bash,linenums=true]
+----
+$ java HelloWorld.java
+Hello world!
+----
+
+For exploring Java in a quick and iterative mode,
+it provides *jshell*, an interactive Read-Evaluate-Print Loop (REPL).
+You can use it for running expressions and statements without defining a surrounding method or class,
+simplifying "Hello World" quite a bit:
+
+[source,bash,linenums=true]
+----
+$ jshell
+
+jshell> System.out.println("Hello World");
+Hello World
+----
+
+Similar to jshell, but quite a bit more fancier is *https://www.jbang.dev/[jbang]*,
+which for instance allows you to easily pull in 3rd party libraries into your single source file Java programs.
+
+== 👷‍♀️ Build Tool
+
+Once you go beyond the basics of manually compiling and running a set of Java classes,
+you'll need a build tool.
+It will help you not only with compiling your code,
+but also with managing dependencies (i.e. 3rd party libraries you are using),
+testing your application, assembling the output artifacts (e.g. a JAR file with your program), and much more.
+There are plug-ins for finding common bugs, auto-formatting your code, etc.
+Commonly used build tool options for Java include Apache Maven, Gradle, and Bazel.
+
+My recommendation is to stick with *Maven* for the beginning;
+it's the most widely used one and in my opinion the easiest to learn.
+While it defines a rather rigid structure for your project,
+that also frees you from having to think about many aspects, which is great in particular when getting started.
+
+== 📝 Editor
+
+Java developers like to fight over their favourite build tools, and it's the same with editors and full-blown integrated development environments (IDEs).
+So whatever I'm going to say here, it's guaranteed a significant number of people won't like it ;)
+My suggestion is to use *VSCode* at the beginning.
+It's a rather light-weight editor, which comes with good Java developer support,
+e.g. for testing and debugging your code.
+It integrates with Maven-based projects very well and has a rich eco-system of plug-ins you can tap into.
+As your needs grow, you'll probably look for an IDE which come with even more advanced functionality, e.g. when it comes to refactoring your code.
+While I'm personally a happy user of the Eclipse IDE,
+most folks tend to use *IntelliJ* these days and it's thus what I'd recommend you too look into.
+It comes with a feature-rich free community edition which will help you a lot will the day-to-day tasks you'll encounter as a Java developer.
+Make sure to spend a few hours to learn the most important keyboard short-cuts,
+it will save you lots of time later on.
+
+== 🐳 Container Image
```

Comment:
Maybe somewhere later? Personally I think frameworks would fit better here (first build an app with ide, build tool, framework, then run it in a container)

Reply:
Yes, good point, will swap them.